zarrs_filesystem/src/lib.rs

```diff
@@ -270,31 +279,70 @@ impl ReadableStorageTraits for FilesystemStore {
 
         let out = byte_ranges
             .map(|byte_range| {
-                let bytes = {
-                    // Seek
-                    match byte_range {
-                        ByteRange::FromStart(offset, _) => file.seek(SeekFrom::Start(offset)),
-                        ByteRange::Suffix(length) => {
-                            file.seek(SeekFrom::End(-(i64::try_from(length).unwrap())))
-                        }
-                    }?;
-
-                    // Read
-                    match byte_range {
-                        ByteRange::FromStart(_, None) => {
-                            let mut buffer = Vec::new();
-                            file.read_to_end(&mut buffer)?;
-                            buffer
-                        }
-                        ByteRange::FromStart(_, Some(length)) | ByteRange::Suffix(length) => {
-                            let length = usize::try_from(length).unwrap();
-                            let mut buffer = vec![0; length];
-                            file.read_exact(&mut buffer)?;
-                            buffer
+                #[cfg(target_os = "linux")]
+                if enable_direct {
+                    let file_size = usize::try_from(file.metadata().unwrap().size()).unwrap(); // FIXME: unwraps
+                    let ps: usize = page_size::get();
+                    let (offset, length) = match byte_range {
+                        ByteRange::FromStart(offset, option_length) => {
+                            let usize_offset = usize::try_from(offset).unwrap();
+                            match option_length {
+                                Some(length) => (usize_offset, usize::try_from(length).unwrap()),
+                                None => (usize_offset, file_size - usize_offset)
+                            }
+                        },
+                        ByteRange::Suffix(offset) => {
+                            let usize_offset = usize::try_from(offset).unwrap();
+                            (file_size - usize_offset, usize_offset)
                         }
+                    };
+                    if length > file_size {
```

Comment:
Should be `offset + length > file_size`, but could just be simplified like this

```suggestion
                    let file_size = file.metadata()?.size();
                    let ps: usize = page_size::get();
                    let range = byte_range.to_range(file_size);
                    let offset = usize::try_from(range.start).unwrap();
                    let length = usize::try_from(range.end - range.start).unwrap();
                    if range.end > file_size {
```